### ISSUE TITLE
contourf with iso-lines that are not closed

### ISSUE THREAD

**Issue by egeerardyn** [matlab2tikz/matlab2tikz#723]
The current implementation of handling filled contour plots in HG2 assumes that all isolines are closed. However, this is not always the case, e.g.

```matlab
N = 20;
[X, Y] = meshgrid(1:N, 1:N);
Z = sqrt(X.^2 ./ Y.^2);
contourf(X+sin(Y), Y, Z, 50, 'EdgeColor', 'k');
matlab2tikz('test.tex','standalone',true);
!pdflatex test.tex
```

This produces the following output:

<img width="1064" alt="screen shot 2015-08-11 at 15 36 47" src="https://cloud.githubusercontent.com/assets/177360/9198787/cf80eb5e-403e-11e5-8406-1d3bf4933293.png">

What we see is that the iso-lines cross the boundary, but are cycled in a straight line by TikZ. Instead, it should be cycled over the bounds of the domain. A possible hurdle that I see is: what part of the boundary: the clockwise or counterclockwise completion?

**Comment by sacraig** [comment]
I'll point out that immediately following the issue I submitted that led you to discover this issue, I have now run into it. I'd actually hazard a guess that in the scientific world, these kinds of contours are actually the most common.

**Comment by okomarov** [comment]
@egeerardyn This example is quite complicated. I would start with something simpler, i.e. when the contours domain covers the whole figure. 

Then there are two cases for open contours:
1 open contour that end on same edge of the box. 
2. open contour that goes from one edge to another

The first case does not involve detecting which direction to cycle, you just need to repeat the first coordinate.

The second can be taken care by using the info on whether it's a hill or a valley and comparing on which side is the next contour of the group.

**Comment by egeerardyn** [comment]
@okomarov you are right this is immediately the full complexity case. In a first step we could cycle over the bounding box around the domain and later on refine that to the actual domain as you said.

During development, your suggestion is a great breakdown to start writing, but for the implementation, we only need the second case (as the first case is also covered by that). My conclusion: let's come up with a contour plot that covers your cases in a simple way.

**Comment by egeerardyn** [comment]
This is just a minor update on what I think is needed to implement this:

 - Determine the boundary of the plot (or first its bounding box, as we do now)
 - For each iso-line:
    * Check if it's non-closed (i.e. first point ~= last point), just to be sure.
    * Check whether it [runs clockwise (CW) or CCW](http://stackoverflow.com/questions/1165647/how-to-determine-if-a-list-of-polygon-points-are-in-clockwise-order)
    * Check on what segment of the boundary the first and last point lie.
    * Complete the iso-line over the boundary in the same direction (CW/CCW) as the rest of the isoline.
